src/TypedFunctionWrapper.php

```diff
@@ -0,0 +1,121 @@
+use ReflectionFunctionAbstract;
+use ReflectionMethod;
+
+class TypedFunctionWrapper extends FunctionWrapper
+{
+    public const FUNCTION_STATUS_HEADER = 'X-Google-Status';
+
+    /** @var callable */
+    protected $function;
+    /** @var ReflectionClass */
+    protected $functionArgClass;
+
+    public function __construct(callable $function)
+    {
+        parent::__construct($function);
+
+        $this->validateFunctionSignature(
+            $this->getFunctionReflection($function)
+        );
+    }
+
+    public function execute(ServerRequestInterface $request): ResponseInterface
+    {
+        try {
+            $argInst = $this->functionArgClass->newInstance();
+        } catch (ReflectionException $e) {
+            throw new LogicException('Function request class cannot be instantiated');
+        }
+        $body = $request->getBody()->getContents();
+
+        try {
+            $argInst->mergeFromJsonString($body);
+        } catch (Exception $e) {
+            throw new ParseError($e->getMessage(), 0, $e);
+        }
+
+        $funcResult = call_user_func($this->function, $argInst);
+
+        if (!method_exists($funcResult, "serializeToJsonString")) {
+            throw new LogicException("Return type must implement 'serializeToJsonString'");
+        }
+        $resultJson = $funcResult->serializeToJsonString();
+
+        return new Response(200, ['content-type' => 'application/cloudevents+json'], $resultJson);
+    }
+
+    private function validateFunctionSignature(
+        ReflectionFunctionAbstract $reflection
+    ) {
+        $parameters = $reflection->getParameters();
+        $parametersCount = count($parameters);
+
+        // Check there is at least one parameter
+        if ($parametersCount === 0) {
+            $this->throwInvalidFunctionException();
+        }
+
+        // Check the first parameter has the proper typehint
+        $type = $parameters[0]->getType();
+        if ($type == null) {
+            $this->throwInvalidFunctionException();
+        }
+
+        for ($i = 1; $i < $parametersCount; $i++) {
+            if (!$parameters[$i]->isOptional()) {
+                throw new LogicException(
+                    'If your function accepts more than one parameter the '
+                        . 'additional parameters must be optional'
+                );
+            }
+        }
+
+        try {
+            $this->functionArgClass = new ReflectionClass($type->getName());
+        } catch (ReflectionException $e) {
+            $name = $type->getName();
+            $message = $e->getMessage();
+            throw new LogicException("Could not find function parameter type $name, error: $message");
+        }
+    }
+
+    private function throwInvalidFunctionException()
+    {
+        throw new LogicException(
+            'Your function must declare exactly one required parameter that has a valid type hint'
+        );
+    }
+}
+
+class ParseError extends Exception
+{
+}
```

Comment:
This seems to be a mistake? [`ParseError`](https://www.php.net/manual/en/class.parseerror.php) is a PHP built-in class and should be used instead of what we have here

Reply:
This error type should be unique as it's only intended to be thrown when the function framework intends to respond with 400 Bad Request. Perhaps `BadRequestError` would be unambiguous?

Reply:
That would make more sense, but it's also bad practice to define more than one class in a class file. This class needs to be placed in its own class file and given proper documentation!

Reply:
Done, moved this to a new file.